Adherent Healing Fixes

Aurum22 opened the pull request:
Fix #25050 

The Adherent mineral bath will now no longer delete your tendril junction and kill you. 
Leaving the mineral bath will now update your icons to match your shiny repaired pieces. 
Embedded objects like shrapnel (but not brig or traitor implants etc) will now be removed by the mineral bath. 
You can now hear the charge pylon filling up your power cell.

Spoke to Gentlefood on discord, should be ok.

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You'll find a README and example file in .\html\changelogs\ for further instructions.

You can also find a template for adding your changelog directly to the PR description here: https://github.com/Baystation12/Baystation12/wiki/Automatic-changelog-generation
-->

:cl: Cronac
bugfix: The Adherent mineral bath will no longer kill Adherent users by deleting their tendril junction.
tweak: The Adherent mineral bath now removed embedded objects like shrapnel, but will not remove implants such as tracking implants etc.
/:cl:

Roland410 commented on code/game/objects/structures/mineral_bath.dm:
```diff
@@ -137,8 +138,13 @@
 				var/obj/item/organ/external/E = thing
 				if(BP_IS_ROBOTIC(E))
+					for(var/obj/implanted_object in E.implants)
+						if(!istype(implanted_object,/obj/item/weapon/implant))	// We don't want to remove REAL implants. Just shrapnel etc.
```
Wouldn't this remove augments as well? Don't recall the path of it.
Also isn't this going to be a tad bit too good? Although this isn't really a code issue

Aurum22 replied:
There's currently no way to remove shrapnel from an Adherent. From some feedback, it may be better to shift it to a chance of happening like the actual healing ticks instead however.

afterthought2 commented:
Can't they just get routine surgery to remove embedded objects?

Roland410 commented:
Apparently adherents were coded in a way that they can't get surgery.
http://prntscr.com/namefo

xales commented on code/game/objects/structures/mineral_bath.dm:
```diff
@@ -125,22 +126,25 @@
 				var/obj/item/organ/external/E = thing
 				if(BP_IS_ROBOTIC(E))
+					for(var/obj/implanted_object in E.implants)
+						if(!istype(implanted_object,/obj/item/weapon/implant || /obj/item/organ/internal/augment) && prob(25))	// We don't want to remove REAL implants. Just shrapnel etc.
```
Does `istype` actually work this way? Never seen this syntax before. May need two `istype` calls.

Aurum22 replied:
I just did a quick search and they always seem separate, so I'll fix that now.

afterthought2 replied:
This will only do the first type check; the `||` is evaluated to the left typepath.

Aurum22 replied:
Just fixed this and did a commit.

comma commented on code/game/objects/structures/mineral_bath.dm:
```diff
@@ -125,22 +126,25 @@
-						repaired_organ = TRUE
-						break
+				if(BP_IS_CRYSTAL(I) && I.damage)
+					I.heal_damage(rand(3,5),rand(3,5))
```
`I.heal_damage(rand(3,5))`, internal organs only take one arg

Aurum22 replied:
Fixed it, squashing now

Aurum22 committed:
Adherent Healing Fixes

The Adherent mineral bath will now no longer delete your tendril junction and kill you.
Leaving the mineral bath will now update your icons to match your shiny repaired pieces.
Embedded objects like shrapnel (but not brig or traitor implants etc) will now be removed by the mineral bath.
You can now hear the charge pylon filling up your power cell.
